File: 10/main.nim
import std/[os,strutils, sugar, strscans, sequtils, sets]


type
  Register = seq[int]
  Screen = string


proc parseInput(): seq[int] =
  var val: int
  for l in paramStr(1).readFile.splitLines:
    if l == "": continue
    if l.scanf("addx $i", val):
      result.add(val)
    else:
      result.add(0)


func get_register(instructions: seq[int]): Register =
  result.add(1)
  for i in instructions:
    if i == 0:
      result.add(result[^1])
    else:
      result.add(@[result[^1], result[^1] + i])


func signal_strength(r: Register, cycles=[20, 60, 100, 140, 180, 220]): int =
  return cycles.map(c => c * r[c-1]).foldl(a + b)


proc draw_screen(r: Register): string =
  result = "#"
  const sprite_pos = [-1, 0, 1].toHashSet
  for idx, i in r:
    if (idx mod 40 - i) in sprite_pos:
      result.add("#")
    else: result.add(".")


proc echo(s: Screen) =
  for idx, i in s:
    if idx mod 40 == 0: stdout.write "\n"
    else: stdout.write i


proc main() =
  let register = parseInput().get_register
  echo register.signal_strength
  echo register.draw_screen


when isMainModule:
  main()

File: 10/main.py
import sys
from typing import List


def read_input():
    with open(sys.argv[1], "r") as f:
        lines = [i.strip() for i in f.readlines()]
    return [0 if i == "noop" else int(i.split()[-1]) for i in lines]


def run_instructions(instructions: List[int]) -> List[int]:
    register = [1]
    for i in instructions:
        if i == 0:
            register.append(register[-1])
        else:
            register.extend([register[-1], register[-1] + i])
    return register


def signal_strength(register: List[int], cycles=[20, 60, 100, 140, 180, 220]) -> int:
    return sum(c * register[c - 1] for c in cycles)


def draw_screen(register: List[int]) -> str:
    screen = ["."] * 240
    sprite_pos = set([-1, 0, 1])
    for idx, x in enumerate(register):
        if idx % 40 - x in sprite_pos:
            screen[idx + 1] = "#"
    return "".join(screen)


def print_screen(screen: str) -> None:
    for idx, i in enumerate(screen):
        if idx % 40 == 0:
            print()
        else:
            print(i, end="")


def main():
    instructions = read_input()
    register = run_instructions(instructions)
    print(signal_strength(register))
    screen = draw_screen(register)
    print_screen(screen)


if __name__ == "__main__":
    main()
